File: backend/src/models/booking.rs
// models/booking.rs

use serde::{Deserialize, Serialize};
use sqlx::prelude::{FromRow, Type};
use chrono::{DateTime, Utc};

#[derive(Debug, Serialize, Deserialize, Type, Clone, Copy, PartialEq, Eq)]
#[sqlx(type_name = "CHAR(3)")]
#[serde(rename_all = "UPPERCASE")]
pub enum BookingStatus {
    WAT,    // Waiting
    RAC,    // Reservation Against Cancellation
    PND,    // Pending
    CNF,    // Confirmed
    CNC,    // Cancelled
}

#[derive(Serialize, Deserialize)]
pub struct CreateBooking {
    pub pnr: i64,
    pub journey_id: i64,
    pub start_station_id: i64,
    pub end_station_id: i64,
    pub price: f32,
    pub txn_id: String,
    pub sid: i64
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Booking {
    pub booking_id: i64,
    pub booking_time: DateTime<Utc>,
    pub booking_status: BookingStatus,
    pub pnr: i64,
    pub journey_id: i64,
    pub start_station_id: i64,
    pub end_station_id: i64,
    pub price: f32,
    pub txn_id: i64,
    pub sid: i64,
}

#[derive(Deserialize)]
pub struct GroupBookingRequest {
    pub group_size: i32,
    pub passenger_data: serde_json::Value, // raw JSON array of passengers
    pub journey_id: i64,
    pub train_id: i64,
    pub start_station_id: i64,
    pub end_station_id: i64,
    pub mode: String,
    pub txn_id: i64,
    pub email: String,
    pub reservation_category: String, // e.g. "SL", "AC3", etc.
}

#[derive(Debug, Serialize)]
pub struct BookingDetail {
    pub pnr: i64,
    pub pass_name: Option<String>,
    pub age: Option<i32>,
    pub sex: Option<String>,
    pub disability: Option<i8>,
    pub booking_id: Option<i64>,
    pub booking_time: Option<chrono::DateTime<Utc>>,
    pub booking_status: Option<String>,
    pub amount: Option<f32>,
    pub txn_id: Option<i64>,
    pub payment_mode: Option<String>,
    pub txn_status: Option<String>,
    pub reservation_status: Option<String>,
    pub reservation_category: Option<String>,
    pub seat_no: Option<i64>,
    pub seat_type: Option<String>,
    pub seat_category: Option<String>,
    pub coach_name: Option<String>,
    pub coach_type: Option<String>,
    pub train_name: Option<String>,
    pub train_type: Option<String>,
    pub journey_id: Option<i64>,
    pub start_time: Option<chrono::DateTime<Utc>>,
    pub end_time: Option<chrono::DateTime<Utc>>,
    pub start_station: Option<String>,
    pub end_station: Option<String>,
}

File: frontend/lib/data/models/booking.dart
class BookingDetail {
  final int pnr;
  final String passName;
  final int age;
  final String sex;
  final bool disability;
  final int bookingId;
  final String bookingTime;
  final String bookingStatus;
  final double amount;
  final int? txnId;
  final String? paymentMode;
  final String? txnStatus;
  final String? reservationStatus;
  final String? reservationCategory;
  final int? seatNo;
  final String? seatType;
  final String? seatCategory;
  final String? coachName;
  final String? coachType;
  final String? trainName;
  final String? trainType;
  final int journeyId;
  final DateTime startTime;
  final DateTime endTime;
  final String? startStation;
  final String? endStation;

  BookingDetail({
    required this.pnr,
    required this.passName,
    required this.age,
    required this.sex,
    required this.disability,
    required this.bookingId,
    required this.bookingTime,
    required this.bookingStatus,
    required this.amount,
    this.txnId,
    this.paymentMode,
    this.txnStatus,
    this.reservationStatus,
    this.reservationCategory,
    this.seatNo,
    this.seatType,
    this.seatCategory,
    this.coachName,
    this.coachType,
    this.trainName,
    this.trainType,
    required this.journeyId,
    required this.startTime,
    required this.endTime,
    this.startStation,
    this.endStation,
  });

  factory BookingDetail.fromJson(Map<String, dynamic> json) => BookingDetail(
    pnr: json['pnr'],
    passName: json['pass_name'],
    age: json['age'],
    sex: json['sex'],
    disability: json['disability'] == 1,
    bookingId: json['booking_id'],
    bookingTime: json['booking_time'],
    bookingStatus: json['booking_status'],
    amount: (json['amount'] as num).toDouble(),
    txnId: json['txn_id'],
    paymentMode: json['payment_mode'],
    txnStatus: json['txn_status'],
    reservationStatus: json['reservation_status'],
    reservationCategory: json['reservation_category'],
    seatNo: json['seat_no'],
    seatType: json['seat_type'],
    seatCategory: json['seat_category'],
    coachName: json['coach_name'],
    coachType: json['coach_type'],
    trainName: json['train_name'],
    trainType: json['train_type'],
    journeyId: json['journey_id'],
    startTime: DateTime.parse(json['start_time']),
    endTime: DateTime.parse(json['end_time']),
    startStation: json['start_station'],
    endStation: json['end_station'],
  );
}


class CancelBookingRequest {
  final int bookingId;
  final double refundAmount;
  final int txnId;

  CancelBookingRequest({
    required this.bookingId,
    required this.refundAmount,
    required this.txnId,
  });

  Map<String, dynamic> toJson() => {
        'booking_id': bookingId,
        'refund_amount': refundAmount,
        'txn_id': txnId,
      };
}
